### PR TITLE
[TF:TRT] Fix Grappler params to enable layout optimizer for the C++ API

### PR DESCRIPTION
**Pull request by tfeher**
This PR sets the VirtualCluster argument while calling the Grappler optimization pass for TF-TRT. Setting the cluster arg is necessary to enable layout optimizer (otherwise layout optimizer returns without changing the graph). This change only affects conversions initiated by the C++ API of TF-TRT #52012.

Tagging @bixia1 for review.

### CODE REVIEW

File: tensorflow/compiler/tf2tensorrt/trt_convert_api.cc [hunk 1]
```diff
@@ -63,8 +64,15 @@ Status RunGrappler(const MetaGraphDef& meta_graph_def,
         "Failed to create grappler item from MetaGraphDef.");
   }
 
+  std::unordered_map<string, tensorflow::DeviceProperties> device_map;
+  string device_name = "/job:localhost/replica:0/task:0/device:GPU:0";
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/trt_convert_api.cc:68]: Why it is safe to assume GPU:0 is used?

**Comment by gbaned** [comment]
@tfeher Can you please check @bixia1's comments and keep us posted ? Thanks!

File: tensorflow/compiler/tf2tensorrt/trt_convert_api.cc (hunk 1 above)
Comment (tfeher) [reply tensorflow/compiler/tf2tensorrt/trt_convert_api.cc]: You are right that we should not have this as a constant. I have changed the cluster initialization, to mirror how the cluster is initialized in the python wrapper
https://github.com/tensorflow/tensorflow/blob/6cda859dbcc86772efc0588ede824ab47ac437f9/tensorflow/python/grappler/cluster_wrapper.cc#L82-L92 which is used when we initiate grappler conversion using the Python API https://github.com/tensorflow/tensorflow/blob/6cda859dbcc86772efc0588ede824ab47ac437f9/tensorflow/python/grappler/tf_optimizer.py#L63

This way both the C++ and the Python API of TF-TRT are calling Grappler with the same cluster settings.

**Review by tfeher — COMMENTED** [review]
Thanks @bixia1 for the review, I have changed the initialization of the grappler cluster.

**Comment by tfeher** [comment]
Added a mutex because we can only provision one cluster per process.

**Review by bixia1 — APPROVED** [review]

File: tensorflow/compiler/tf2tensorrt/trt_convert_api.cc [hunk 2]
```diff
@@ -42,10 +46,29 @@ namespace tensorflow {
 namespace tensorrt {
 namespace {
 
+struct ClusterDestroyer {
+  void operator()(grappler::Cluster* c) {
+    if (c) c->Shutdown();
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/trt_convert_api.cc:51]: this returns a Status. How would you want to use the return value?

**Review by bixia1 — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/tf2tensorrt/trt_convert_api.cc (hunk 2 above)
Comment (tfeher) [reply tensorflow/compiler/tf2tensorrt/trt_convert_api.cc]: Sorry I have missed it. If an error occurs, then we should return a conversion error. But this is not possible with the current ClusterDestroyer, so I have abandoned this functor, and call shutdown directly. Let me know if you have a better suggestion.

**Review by tfeher — COMMENTED** [review]
Thanks @bixia for pointing out the issue, it is fixed now.

**Review by bixia1 — APPROVED** [review]

**Comment by gbaned** [comment]
@tfeher  Can you please address Ubuntu Sanity errors? Thanks!

**Comment by tfeher** [comment]
The logs says `"ERROR: An error occurred during the fetch of repository 'rules_proto':"` This is not related to my changes.

**Comment by tfeher** [comment]
The changes are already included in master, in commit https://github.com/tensorflow/tensorflow/commit/1d5124678decc63d3b64f2598b91a61169748c3e. Closing the PR.

**Pull request closed without merging**